### PR TITLE
Only try to backfill event if we haven't tried before recently

### PR DESCRIPTION
**Pull request by MadLittleMods**
Only try to backfill event if we haven't tried before recently (exponential backoff). No need to keep trying the same backfill point that fails over and over.

Fix https://github.com/matrix-org/synapse/issues/13622
Fix https://github.com/matrix-org/synapse/issues/8451

Follow-up to https://github.com/matrix-org/synapse/pull/13589

Part of https://github.com/matrix-org/synapse/issues/13356

---

The `get_oldest_event_ids_with_depth_in_room` tests are using this DAG layout:

```mermaid
flowchart BT

    3 --> A
    subgraph backfillPointsA
        b1:::not_backfilled_yet
        b2:::not_backfilled_yet
        b3:::not_backfilled_yet
    end

    subgraph main shoot
        5 --> 4 --> 3 --> 2:::not_backfilled_yet --> 1:::not_backfilled_yet
    end

    4 --> B
    subgraph backfillPointsB
        b4:::not_backfilled_yet
        b5:::not_backfilled_yet
        b6:::not_backfilled_yet
    end


    A --> b1 --> 2
    A --> b2 --> 2
    A --> b3 --> 2

    B --> b4 --> 3
    B --> b5 --> 3
    B --> b6 --> 3


    classDef not_backfilled_yet opacity: 0.75,stroke-dasharray: 5 5;
```

### Dev notes

```
SYNAPSE_POSTGRES=1 SYNAPSE_TEST_LOG_LEVEL=DEBUG python -m twisted.trial tests.storage.test_event_federation.EventFederationWorkerStoreTestCase.test_get_backfill_points_in_room

SYNAPSE_TEST_LOG_LEVEL=DEBUG python -m twisted.trial tests.storage.test_event_federation.EventFederationWorkerStoreTestCase.test_get_backfill_points_in_room
```

---

<details>
<summary>Previous different graph</summary>

```mermaid
flowchart BT

    3 --> A
    subgraph backfillPointsA
        b1:::not_backfilled_yet
        b2:::not_backfilled_yet
        b3:::not_backfilled_yet
    end
    
    subgraph main shoot
        5 --> 4 --> 3 --> 2:::not_backfilled_yet --> 1:::not_backfilled_yet
    end

    4 --> B
    subgraph backfillPointsB
        b4:::not_backfilled_yet
        b5:::not_backfilled_yet
        b6:::not_backfilled_yet
    end


    A --> b1 --> 2
    A --> b2 --> 2
    A --> b3 --> 2

    B --> b4 ------> 2
    B --> b5 ------> 2
    B --> b6 ------> 2


    classDef not_backfilled_yet opacity: 0.75,stroke-dasharray: 5 5;

    %% alignment links
    B --- alignment1[ ]
    %% make the alignment links/nodes invisible
    style alignment1 visibility: hidden,color:transparent;
    linkStyle 18 stroke-width:2px,fill:none,stroke:none;
```

</details>

----

 - Complement Docker SQLite version: `3.34.1` `sqlite3.sqlite_version_info=(3, 34, 1)`
 - My local `sqlite3 --version` on macOS machine -> `3.37.0 2021-12-09 01:34:53 9ff244ce0739f8ee52a3e9671adb4ee54c83c640b02e3f9d185fd2f9a179aapl`

---

```
‹
{
´
`
,
⹁
```

### Todo

 - [x] Do same `depth` stuff to `get_insertion_event_backward_extremities_in_room`
    - [x] Add tests
 - [x] Remove the `depth` filter from `_maybe_backfill_inner` itself because it's already done in the queries
 - [ ] Future PR: Do we need to consider it a failed backfill attempt when we ask for backfill for a certain event and don't receive it in the response? -> https://github.com/matrix-org/synapse/issues/13700
    - This isn't really a direct concern of this PR. We can iterate on this later.
 - [x] Figure out how to handle `power` support in SQLite, https://github.com/matrix-org/synapse/pull/13635#discussion_r956543645
 - [x] Fix `sqlite3.OperationalError: no such column: failed_backfill_attempt_info`, https://github.com/matrix-org/synapse/pull/13635#discussion_r959929330
 - [ ] Future PR: Add a `limit` to both `get_oldest_event_ids_with_depth_in_room` and `get_insertion_event_backward_extremities_in_room` so we only try as much as we need to 

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [ ] ~~Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)~~
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/storage/databases/main/event_federation.py [hunk 1]
```diff
@@ -735,34 +744,69 @@ async def get_oldest_event_ids_with_depth_in_room(
...
+                     * We only want backwards extremities that are older than or at
+                     * the same position of the given `current_depth` (where older
+                     * means less than a given depth).
+                     */
+                    AND MAX(event.depth) <= current_depth
+                    /**
+                     * Exponential back-off (up to the upper bound) so we don't retry the 
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc
+                     */
+                    AND ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least(2^backfill_attempt_info.num_attempts * ?, ? /* upper bound */)
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:790]: Is implementing the exponential back-off lookup in the SQL query acceptable?

I could imagine changing the schema `event_backfill_attempts` to include the time it should retry at but then that bakes in the time to the database and the config could change.

File: synapse/storage/databases/main/event_federation.py [hunk 2]
```diff
@@ -71,6 +72,12 @@
 
 logger = logging.getLogger(__name__)
 
+BACKFILL_EVENT_BACKOFF_UPPER_BOUND_SECONDS: int = int(
+    datetime.timedelta(days=7).total_seconds()
+)
+BACKFILL_EVENT_EXPONENTIAL_BACKOFF_STEP_SECONDS: int = int(
+    datetime.timedelta(hours=1).total_seconds()
+)
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:80]: New config?

File: synapse/storage/databases/main/event_federation.py (hunk 1 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:795]: Commented on https://github.com/matrix-org/synapse/pull/13589#discussion_r955588328 to make sure we're going down the right path

File: synapse/storage/databases/main/event_federation.py [hunk 3]
```diff
@@ -735,34 +744,78 @@ async def get_oldest_event_ids_with_depth_in_room(
...
+                    /**
+                     * Exponential back-off (up to the upper bound) so we don't retry the 
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc
+                     */
+                    AND (
+                        backfill_attempt_info IS NULL
+                        OR ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least(2^backfill_attempt_info.num_attempts * ?, ? /* upper bound */)
+                    )
+                /* TODO: Why */
+                GROUP BY backward_extrem.event_id
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:800]: I'm not sure why we had this before? There is only ever one backward extremity with a given `event_id`.

Removing it and the query seems to behave just as good as before.

File: synapse/storage/databases/main/event_federation.py [hunk 4]
```diff
@@ -735,34 +744,78 @@ async def get_oldest_event_ids_with_depth_in_room(
...
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_backfill_attempts as backfill_attempt_info
+                ON backfill_attempt_info.event_id = backward_extrem.event_id
+                WHERE
+                    backward_extrem.room_id = ?
+                    /* We only care about normal events because TODO: why */
+                    AND edge.is_state is ? /* False */
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:779]: Why do we only care about normal events here?

It seems just as valid to backfill from a state event which is also part of the timeline.

File: synapse/storage/databases/main/event_federation.py (hunk 2 above)
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:80]: New config?

I'm happy to leave it as constants for now until someone has a need to customize.

Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:80]: Any opinions on these values?

I chose 7 days because trying again next week seemed reasonable if someone's server was offline and then they got it working again.

This means it takes 8 attempts to get to the upper bound (`2^7 = 128`, `2^8 = 256`)

2hr, 4hr, 8hr, 16hr, 32hr, 64hr, 128hr, (capped at 168hr from now on)

File: synapse/storage/databases/main/event_federation.py [hunk 5]
```diff
@@ -714,77 +722,152 @@ def _get_auth_chain_difference_txn(
...
+                     * <oldest-in-time> [0]<--[1]▼<--[2]<--[3]<--[4] <newest-in-time>
+                     */
+                    AND event.depth <= ? /* current_depth */
+                    /**
+                     * Exponential back-off (up to the upper bound) so we don't retry the
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc
+                     */
+                    AND (
+                        backfill_attempt_info IS NULL
+                        OR ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least(2^backfill_attempt_info.num_attempts * ? /* step */, ? /* upper bound */)
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:811]: One problem with doing the calculation in SQL is that SQLite doesn't like the `^` power operator, `sqlite3.OperationalError: unrecognized token: "^"`

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:811]: via https://stackoverflow.com/a/67105253/796832, it looks like SQLite does have `power(X, Y)` available but it was added in [SQLite v3.35.0](https://www.sqlite.org/changes.html#version_3_35_0) (2021-03-12) and only as a compile-time option  `-DSQLITE_ENABLE_MATH_FUNCTIONS` 😩

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:811]: via https://stackoverflow.com/a/55264405/796832, it's possible to create a user-defined function from Python and looks like we already do it so this might be viable:

https://github.com/matrix-org/synapse/blob/c4e29b6908ac8ae57b5e9a3e7662ad638b61e94a/synapse/storage/engines/sqlite.py#L98

Comment (reivilibre) [reply synapse/storage/databases/main/event_federation.py]: You can use the left shift operator: `2 << (n - 1)` instead of `pow(2, n)`. It's in Debian oldstable's version of SQLite, so it should be supported. :)

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py]: Ahh, great thinking @reivilibre! 😇

For onlookers, only works with powers of `2` because left shift is a binary operation aka working with base-2.

Added in SQLite 2.0.3 (20001-10-13), https://www.sqlite.org/changes.html#version_2_0_3

File: synapse/handlers/federation.py [hunk 6]
```diff
@@ -276,10 +278,6 @@ async def _maybe_backfill_inner(
             insertion_events_to_be_backfilled,
         )
 
-        if not backwards_extremities and not insertion_events_to_be_backfilled:
-            logger.debug("Not backfilling as no extremeties found.")
-            return False
-
         # we now have a list of potential places to backpaginate from. We prefer to
         # start with the most recent (ie, max depth), so let's sort the list.
         sorted_backfill_points: List[_BackfillPoint] = sorted(
```
Comment (MadLittleMods) [inline synapse/handlers/federation.py:283]: Can we do anything smarter with two already sorted lists?

File: synapse/storage/databases/main/event_federation.py [hunk 7]
```diff
@@ -714,77 +722,157 @@ def _get_auth_chain_difference_txn(
...
+                     *
+                     * We use `2 << (n - 1)` as a power of 2 equivalent for compatibility
+                     * with SQLite. The left shift equivalent only works with powers of 2
+                     * because left shift is a binary operation (base-2). Otherwise, we
+                     * would use `power(2, n)` or the power operator, `2^n`.
+                     */
+                    AND (
+                        backfill_attempt_info IS NULL
+                        OR ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least((2 << (backfill_attempt_info.num_attempts - 1)) * ? /* step */, ? /* upper bound */)
+                    )
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:817]: The query currently results in `sqlite3.OperationalError: no such column: backfill_attempt_info` but only for SQLite 🤔

Not sure why the left joined table alias is acting like this.

File: synapse/storage/databases/main/event_federation.py [hunk 8]
```diff
@@ -714,77 +722,157 @@ def _get_auth_chain_difference_txn(
...
+               we drop those extremities.
 
         Args:
             room_id: Room where we want to find the oldest events
+            current_depth: The depth at the users current scrollback position because
+                we only care about finding events older than the given
+                `current_depth` when scrolling and paginating backwards.
 
         Returns:
             List of (event_id, depth) tuples
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:755]: Comment about how it's a sorted by depth list

File: synapse/storage/databases/main/event_federation.py (hunk 7 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py]: It's reproducible with just the following snippet ([dbfiddle](https://www.db-fiddle.com/f/eofX81DQdhiPggK22tMDME/0)). SQLite doesn't like the joined table check against `NULL`. Happens with and without the table alias.

`Query Error: Error: SQLITE_ERROR: no such column: failed_backfill_attempt_info`
```
SELECT event_id FROM events AS event
LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info USING (event_id)
WHERE
    event_id = '$event1'
    AND failed_backfill_attempt_info IS NULL;
```

Setup:
```
CREATE TABLE IF NOT EXISTS events(
    event_id TEXT NOT NULL,
    PRIMARY KEY (event_id)
);

CREATE TABLE IF NOT EXISTS event_failed_backfill_attempts(
    event_id TEXT NOT NULL,
    PRIMARY KEY (event_id)
);

INSERT INTO events VALUES ('$event1'), ('$event2'), ('$event3');
INSERT INTO event_failed_backfill_attempts VALUES ('$event1');
```

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py]: Looks like it works if we just use one of the columns from the table instead of the table itself:

```sql
SELECT event_id FROM events AS event
LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info USING (event_id)
WHERE
    event_id = '$event1'
    AND failed_backfill_attempt_info.event_id IS NULL;
```

Solved for now but curious about docs around this ✅

File: synapse/storage/databases/main/event_federation.py [hunk 9]
```diff
@@ -714,96 +722,227 @@ def _get_auth_chain_difference_txn(
...
                 /* Get the depth of the insertion event from the events table */
-                INNER JOIN events AS e USING (event_id)
-                WHERE b.room_id = ?
-                GROUP BY b.event_id
+                INNER JOIN events AS event USING (event_id)
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info USING (room_id, event_id)
```
Comment (MadLittleMods) [inline synapse/storage/databases/main/event_federation.py:903]: Error as seen in the failing CI, https://github.com/matrix-org/synapse/runs/8123978698?check_suite_focus=true#step:4:4915, https://github.com/matrix-org/synapse/runs/8123978792?check_suite_focus=true#step:4:18612

```
 psycopg2.errors.AmbiguousColumn: common column name "room_id" appears more than once in left table
```

Not sure why I can't reproduce locally 🤔

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py]: Solved by specifying specific columns instead of `using`

```sql
LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info
ON
    failed_backfill_attempt_info.room_id = insertion_event_extremity.room_id
    AND failed_backfill_attempt_info.event_id = insertion_event_extremity.event_id
```

File: tests/storage/test_event_federation.py [hunk 10]
```diff
@@ -571,11 +587,449 @@ def prev_event_format(prev_event_id: str) -> Union[Tuple[str, dict], str]:
...
+        self.reactor.advance(datetime.timedelta(hours=20).total_seconds())
+
+        # Try at "A" again after we advanced enough time and we should see "b3" again
+        backfill_points = self.get_success(
+            self.store.get_backfill_points_in_room(room_id, depth_map["A"])
+        )
+        backfill_event_ids = [backfill_point[0] for backfill_point in backfill_points]
+        self.assertListEqual(backfill_event_ids, ["b3", "b2", "b1"])
+
+    def _setup_room_for_insertion_backfill_tests(self) -> _BackfillSetupInfo:
```
Comment (MadLittleMods) [inline tests/storage/test_event_federation.py:848]: The next set of 3 tests for `get_insertion_event_backward_extremities_in_room` are very similar to the 3 tests added for `get_backfill_points_in_room`.

We could probably re-use and generalize these into a test matrix but the everything would be less straightforward. I'm leaning towards leaving this as-is vs complicating it.

**Review by reivilibre — COMMENTED** [review]
some comments, but not a full review

File: synapse/storage/databases/main/event_federation.py [hunk 11]
```diff
@@ -714,96 +722,231 @@ def _get_auth_chain_difference_txn(
...
-        """Gets the oldest events(backwards extremities) in the room along with the
-        aproximate depth.
-
-        We use this function so that we can compare and see if someones current
-        depth at their current scrollback is within pagination range of the
-        event extremeties. If the current depth is close to the depth of given
-        oldest event, we can trigger a backfill.
+        """
+        Gets the oldest events(backwards extremities) in the room along with the
+        aproximate depth (sorted by depth descending).
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:732]: `(sorted by deepest events first)` I think this is ambiguous. Deepest from what point of view.

But also agreed that trying to think about descending/ascending can take a second.

I've updated to:

```
Sorted by depth, highest to lowest (descending) so the closest events to the `current_depth` are first in the list.
```

File: synapse/storage/databases/main/event_federation.py [hunk 12]
```diff
@@ -714,96 +722,231 @@ def _get_auth_chain_difference_txn(
...
+
+        We use this function so that we can compare and see if someones
+        `current_depth` at their current scrollback is within pagination range
+        of the event extremeties. If the `current_depth` is close to the depth
+        of given oldest event, we can trigger a backfill.
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:737]: Correct, the person represented by the client 👍

Updated ✅ 

---

Aside: We have a lot of `extremeties` typos in the codebase including table names 😬

File: synapse/storage/databases/main/event_federation.py [hunk 13]
```diff
@@ -714,96 +722,231 @@ def _get_auth_chain_difference_txn(
...
+                INNER JOIN event_backward_extremities AS backward_extrem
+                ON edge.prev_event_id = backward_extrem.event_id
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_failed_backfill_attempts AS failed_backfill_attempt_info
+                ON
+                    failed_backfill_attempt_info.room_id = backward_extrem.room_id
+                    AND failed_backfill_attempt_info.event_id = backward_extrem.event_id
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:791]: See https://github.com/matrix-org/synapse/pull/13701 which summarizes some of the tribal knowledge around this.

tldr; it's good to guard against `event_id` not be unique globally. Use `(room_id, event_id)`

File: synapse/storage/databases/main/event_federation.py [hunk 14]
```diff
@@ -714,96 +722,231 @@ def _get_auth_chain_difference_txn(
...
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc.
+                     *
+                     * We use `2 << (n - 1)` as a power of 2 equivalent for compatibility
+                     * with SQLite. The left shift equivalent only works with powers of 2
+                     * because left shift is a binary operation (base-2). Otherwise, we
+                     * would use `power(2, n)` or the power operator, `2^n`.
+                     */
+                    AND (
+                        failed_backfill_attempt_info.event_id IS NULL
+                        OR ? /* current_time */ >= failed_backfill_attempt_info.last_attempt_ts + /*least*/%s((2 << (failed_backfill_attempt_info.num_attempts - 1)) * ? /* step */, ? /* upper bound */)
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:818]: No worries, it made sense in my head doing the quick math too 😵

Just to sanity check, looks good ✅:

 - `2^0 = 1` = `1 << 0 = 1`
 - `2^1 = 2` = `1 << 1 = 2`
 - `2^2 = 4` = `1 << 2 = 4`
 - `2^8 = 256` = `1 << 8 = 256`

File: synapse/storage/databases/main/event_federation.py (hunk 1 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py]: Based on no apprehension to include that info in the database table from https://github.com/matrix-org/synapse/pull/13589, I think we're happy to do this ⏩

File: synapse/handlers/federation.py (hunk 6 above)
Comment (squahtx) [reply synapse/handlers/federation.py:252]: There's [`heapq.merge`](https://docs.python.org/3/library/heapq.html#heapq.merge) in the stdlib which runs in linear time. Though it probably doesn't make much difference for any reasonable number of extremities.

File: synapse/storage/databases/main/event_federation.py [hunk 15]
```diff
@@ -715,96 +723,238 @@ def _get_auth_chain_difference_txn(
...
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_failed_pull_attempts AS failed_backfill_attempt_info
+                ON
+                    failed_backfill_attempt_info.room_id = backward_extrem.room_id
+                    AND failed_backfill_attempt_info.event_id = backward_extrem.event_id
+                WHERE
+                    backward_extrem.room_id = ?
+                    /* We only care about non-state events because TODO: why */
```
Comment (erikjohnston) [inline synapse/storage/databases/main/event_federation.py:798]: "state event edges" they're an old holdover and we should rip them out.

File: synapse/storage/databases/main/event_federation.py [hunk 16]
```diff
@@ -715,96 +723,238 @@ def _get_auth_chain_difference_txn(
...
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc.
+                     *
+                     * We use `1 << n` as a power of 2 equivalent for compatibility
+                     * with older SQLites. The left shift equivalent only works with
+                     * powers of 2 because left shift is a binary operation (base-2).
+                     * Otherwise, we would use `power(2, n)` or the power operator, `2^n`.
+                     */
+                    AND (
+                        failed_backfill_attempt_info.event_id IS NULL
+                        OR ? /* current_time */ >= failed_backfill_attempt_info.last_attempt_ts + /*least*/%s((1 << failed_backfill_attempt_info.num_attempts) * ? /* step */, ? /* upper bound */)
```
Comment (erikjohnston) [inline synapse/storage/databases/main/event_federation.py:822]: I wonder if its cleaner if we just make `failed_backfill_attempt_info` include a column for "next time to retry"?

File: synapse/handlers/federation.py [hunk 17]
```diff
@@ -269,6 +267,29 @@ async def _maybe_backfill_inner(
             sorted_backfill_points,
         )
 
+        # If we have no backfill points lower than the `current_depth` then
+        # either we can a) bail or b) still attempt to backfill. We opt to try
+        # backfilling anyway just in case we do get relevant events.
```
Comment (erikjohnston) [inline synapse/handlers/federation.py:272]: This feels like a bit of an orthogonal change? I'm not sure whether we really want to do this

File: synapse/storage/databases/main/event_federation.py (hunk 16 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:822]: This was mentioned in https://github.com/matrix-org/synapse/pull/13589#discussion_r955588328 and cross-referenced here at https://github.com/matrix-org/synapse/pull/13635#discussion_r955585744

Based on no feedback about including the calculation in the SQL query I went with it. It also allows us to tune the timing config and not have to do a bunch of maintenance to update the database rows.

File: synapse/storage/databases/main/event_federation.py (hunk 15 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:798]: Previous discussion thread: https://github.com/matrix-org/synapse/pull/13635#discussion_r956534576

"state event edges" doesn't elucidate anything for me? What were they before? Why did we care? I don't see any reference to them in the current code base.

Can we remove this extra condition in the SQL query?

File: synapse/storage/databases/main/event_federation.py (hunk 4 above)
Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py]: Another discussion thread: https://github.com/matrix-org/synapse/pull/13635#discussion_r977864219

File: synapse/handlers/federation.py (hunk 17 above)
Comment (MadLittleMods) [reply synapse/handlers/federation.py:272]: Fair enough, filtering by `current_depth` in the query itself is an extra fix that we can be split out to another PR. And I can include the additional `limit` change I want to make in that future PR as well

For reference, the change is targeting not pulling out thousands of backfill points from the database every time we need to backfill. We only need to pull out a 20 relevant ones at most since we just use 5 of them.

And there is no functional change ([previous code](https://github.com/matrix-org/synapse/blob/f7c89c44c540d82c4e82a232e6ae1084b7b94e33/synapse/handlers/federation.py#L295-L320))

Comment (MadLittleMods) [reply synapse/handlers/federation.py]: Removed from this PR and split off to https://github.com/matrix-org/synapse/pull/13879 ⏩

File: synapse/storage/databases/main/event_federation.py (hunk 15 above)
Comment (richvdh) [reply synapse/storage/databases/main/event_federation.py:778]: related: https://github.com/matrix-org/synapse/issues/12892

we used to use `event_edges` for two different sorts of "edges" (the current event DAG, but also a link to the previous state, for state events). They can be distinguished by `is_state`.

https://github.com/matrix-org/synapse/pull/12893 removes the rows with `is_state=True`, but unfortunately, because it is a background update, it's not necessarily safe to assume that it will have completed.

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:778]: *chefs kiss

Added this context to the code here 🎉

Thanks for the info @erikjohnston and @richvdh!

**Review by erikjohnston — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @reivilibre and @erikjohnston! Also thanks to @squahtx and @richvdh for the tips! 🦆

File: synapse/storage/databases/main/event_federation.py [hunk 18]
```diff
@@ -715,96 +723,189 @@ def _get_auth_chain_difference_txn(
...
```
Comment (DMRobertson) [inline synapse/storage/databases/main/event_federation.py:798]: Two things here.

1. If num_attempts is 10 then (1 << 10) ** 3600000 overflows a 32bit signed int and postgres errors. See https://github.com/matrix-org/synapse/issues/13929 for the investigation.
2. (Minor): the name "step" for the constant 3600000 (1 hour in ms) suggests that the first backoff period should be 1 hour: but as you noted above the first backoff period is 2 horus.

I'm attempting to PR a fix for (1).

Comment (MadLittleMods) [reply synapse/storage/databases/main/event_federation.py:798]: PR to fix this: https://github.com/matrix-org/synapse/pull/13936
